datajoint/blob.py

```diff
@@ -34,286 +37,384 @@
+
+        blob += np.array([type_id, is_complex], dtype=np.uint32).tobytes()
+        if type_names[type_id] == 'mxVOID_CLASS':  # array of dtype('O')
+            blob += b"".join(len_u64(it) + it for it in (self.pack_blob(e) for e in array.flatten(order="F")))
+            self.set_dj0()  # not supported by original mym
+        elif type_names[type_id] == 'mxCHAR_CLASS':  # array of dtype('c')
+            blob += array.view(np.uint8).astype(np.uint16).tobytes()  # convert to 16-bit chars for MATLAB
+        else:  # numeric arrays
+            if array.ndim == 0:  # not supported by original mym
+                self.set_dj0()
+            blob += array.tobytes(order="F")
+            if is_complex:
+                blob += imaginary.tobytes(order="F")
+        return blob
 
-    def read_string(self, advance=True):
-        """
-        Read a string terminated by null byte '\0'. The returned string
-        object is ASCII decoded, and will not include the terminating null byte.
+    def read_recarray(self):
         """
-        target = self._blob.find(b'\0', self.pos)
-        assert target >= self._pos
-        data = self._blob[self._pos:target]
-        if advance:
-            self._pos = target + 1
-        return data.decode('ascii')
-
-    def read_value(self, dtype='uint64', count=1, advance=True):
+        Serialize an np.ndarray with fields, including recarrays
         """
-        Read one or more scalars of the indicated dtype. Count specifies the number of
-        scalars to be read in.
-        """
-        data = np.frombuffer(self._blob, dtype=dtype, count=count, offset=self.pos)
-        if advance:
-            # probably the same thing as data.nbytes * 8
-            self._pos += data.dtype.itemsize * data.size
-        if count == 1:
-            data = data[0]
+        n_fields = self.read_value('uint32')
+        if not n_fields:
+            return np.array(None)  # empty array
+        field_names = [self.read_zero_terminated_string() for _ in range(n_fields)]
+        arrays = [self.read_blob() for _ in range(n_fields)]
+        rec = np.empty(arrays[0].shape, np.dtype([(f, t.dtype) for f, t in zip(field_names, arrays)]))
+        for f, t in zip(field_names, arrays):
+            rec[f] = t
+        return rec.view(np.recarray)
+
+    def pack_recarray(self, array):
+        """ Serialize a Matlab struct array """
+        return (b"F" + len_u32(array) +  # number of fields
+                '\0'.join(array.dtype.names).encode() + b"\0" +  # field names
+                b"".join(self.pack_recarray(array[f]) if array[f].dtype.fields else self.pack_array(array[f])
+                         for f in array.dtype.names))
+
+    def read_sparse_array(self):
+        raise DataJointError('datajoint-python does not yet support sparse arrays. Issue (#590)')
+
+    def read_decimal(self):
+        return Decimal(self.read_string())
+
+    @staticmethod
```

Comment:
Was there not a lighter way to pack a decimal e.g. 2 integers? Curious why string was utilized.